### ISSUE TITLE
Ability to suspend an argo application

### ISSUE THREAD

**Issue by andye2004** [argoproj/argo-cd#3039]
Firstly, thanks for a great tool. It's made life much easier for us!

There are times where I want to 'suspend' an application. That is, I want to temporarily delete it from the cluster and then re-add it. Currently I have to delete the argo application completely and then re-add it into argo. It would be useful if there was a feature where I could suspend it in argo and that would delete the app in the cluster BUT keep the argo application definition ready to re-sync.

**Comment by jannfis** [comment]
Hm. You could have your application definition residing in a dedicated definition modeling the [app of apps patterns](https://argoproj.github.io/argo-cd/operator-manual/declarative-setup/#app-of-apps) with auto-sync disabled on the parent. 

Then you could cascade-delete the application in question ("child app"), and if you want it back, you just re-sync the parent application.

**Comment by andye2004** [comment]
@jannfis, thanks for the follow-up. I'll have a think on this for a few days and we'll maybe have a discussion internally re your suggestion. Doesn't feel quite right though - I'll get back to you.

**Comment by dcharbonnier** [comment]
This "status" of disabled application could be really useful for development.
If I have two projects using the pattern app of apps :
project1 : app1, app2
project2 : app1, app2, app3

I would be able (on my development machine) to stop working on project1 and start working on project2 with 2 clicks (or one commit)
I don't see a use-case for a production environment.

**Comment by hilenko** [comment]
+1

**Comment by jthunderbird** [comment]
+1

**Comment by davem-git** [comment]
+1

**Comment by nikcollins** [comment]
+1

**Comment by br95187** [comment]
+1

**Comment by armenr** [comment]
+1

**Comment by vmironichev-nix** [comment]
+1

**Comment by tanalam2411** [comment]
+1

**Comment by achetronic** [comment]
+1

**Comment by Cave-Johnson** [comment]
+1

**Comment by JBogossianKH** [comment]
+1 ❤️

**Comment by Va1** [comment]
+1

**Comment by kschjeld** [comment]
+1
Much better than commenting out contents of YAML-file if you want to temporarily disable

**Comment by zoop-btc** [comment]
+1 would be useful

**Comment by Mahesh-Gunda-Maersk** [comment]
+1 very useful.

**Comment by filipdadgar** [comment]
+1, would love to see this feature!

**Comment by elendis** [comment]
+1

**Comment by zuberahmed1987** [comment]
+1

**Comment by medoedoff** [comment]
+1

**Comment by mat-work** [comment]
+1

**Comment by agusserraetendo** [comment]
+1

**Comment by eaceaser** [comment]
+1

**Comment by crenshaw-dev** [comment]
I want to make sure we're all +1'ing the same feature request. 

Please react to indicate which feature you want:

* ❤️ Delete all resources in the app
* 🚀 Scale down all scalable resource in the app
* 👀 Something else (please comment, too)

Please no more +1s without describing your use case. The enthusiasm is appreciated, but keeping the thread tidy is good too. 😄

**Comment by canghai118** [comment]
> I want to make sure we're all +1'ing the same feature request.
> 
> Please react to indicate which feature you want:
> 
> * ❤️ Delete all resources in the app
> * 🚀 Scale down all scalable resource in the app
> * 👀 Something else (please comment, too)
> 
> Please no more +1s without describing your use case. The enthusiasm is appreciated, but keeping the thread tidy is good too. 😄

```
* ❤️ Delete all resources in the app (It's better to have an option to keep the PVC or PV, so the entire applicaiton state can be resumed)
```

**Comment by jeacott1** [comment]
@canghai118 for me both options would be useful actually,
I want to be able to minimise resource usage without trashing configMaps etc so a Scale down option would also be very useful for things that can scale down.
I would also like to sometimes "Delete all resources in the app" without having to delete the app itself.

**Comment by jannfis** [comment]
I think this should be possible meanwhile, although it's not exposed through the UI or CLI (yet): https://argo-cd.readthedocs.io/en/stable/user-guide/skip_reconcile/

This should effectively implement the requested feature?

**Comment by crenshaw-dev** [comment]
@jannfis I think the issue title is too vague. Seems like most folks on the thread are interested in a per-app scale-down mechanism.

**Comment by crenshaw-dev** [comment]
Ooh. With the new actions enhancements in 2.8, we should do something like this:

```lua
-- Loop over resources from the Application status. Scale down Deployments and StatefulSets.

actions = {}

if obj.status == nil then
    return actions
end

for _, resource in ipairs(obj.status.resources or {}) do
    if resource.health.status ~= "Missing" and (resource.kind == "Deployment" or resource.kind == "StatefulSet") then
        action = {}
        action.operation = "patch"
        action.resource = {}
        if resource.group == nil or resource.group == "" then
          action.resource.apiVersion = resource.version
        else
          action.resource.apiVersion = resource.group .. "/" .. resource.version
        end
        action.resource.kind = resource.kind
        action.resource.metadata = {}
        action.resource.metadata.name = resource.name
        action.resource.metadata.namespace = resource.namespace
        action.resource.spec = {}
        action.resource.spec.replicas = 0
        table.insert(actions, action)
    end
end

return actions
```

**Comment by sav116** [comment]
+ 1

**Comment by crenshaw-dev** [comment]
Is anyone up for trying to implement the actions-based solution? I'd be happy to help.

**Comment by crenshaw-dev** [comment]
For the two folks who answered that they want "delete all resources in the app," let's track that feature request here: https://github.com/argoproj/argo-cd/issues/6079

**Comment by jnkxcel** [comment]
For my use case I work at a utility and they heavy hand the password, rotation is a pain, I used to just kubectl scale --replicas=0. But with Argo control it self heals. Which I love but now I have to remove my deploy, update password and then redeploy. Not huge lift but cumbersome for prod facing apps. This is specific to K8s plugin external-dns connecting to infoblox. I manage the external-dns plugin via an app of apps deploy so I just have to remove one line entry and argo sync so again not huge lift. 

Just would be so much nicer to have scale down feature.

**Comment by crenshaw-dev** [comment]
Sounds like in your case you'd need both to both scale down and disable self-heal.

**Comment by MioOgbeni** [comment]
Hi, how is this issue going? It would also be awesome to be able to set scale down of scalable resources by on/off office hours, or by scale down/up cron. It could be configurable by completely new ArgoCD resource.

**Comment by MaxWinterstein** [comment]
> Hi, how is this issue going? It would also be awesome to be able to set scale down of scalable resources by on/off office hours, or by scale down/up cron. It could be configurable by completely new ArgoCD resource or directly in Application resource.

@MioOgbeni you might wanna have a look at https://keda.sh/ and its [cron scaler](https://keda.sh/docs/2.13/scalers/cron/).

**Comment by jnkxcel** [comment]
I found workaround if you need temp solution, just make application manual sync, select the deployment object, select to edit it, in replica key:value shift value to 0. This will effectively shift application to 0 pod/replicas. Update code then refresh and sync and you kind of temp scale down method.

**Comment by evanstucker-hates-2fa** [comment]
How to disable applications, lol.

Ensure that your ApplicationSet `syncPolicy.automated.prune: true`, then:
```
$ git status
On branch main
Your branch is up to date with 'origin/main'.

Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
	renamed:    applications/matomo/config.json -> applications-disabled/matomo/config.json
	renamed:    applications/matomo/values.yaml -> applications-disabled/matomo/values.yaml
```
Note that this may cause gaps in your git history if you move files back and forth. Maybe a smarter person would use symlinks? This is just a joke/hack, but it works. Let's get a real feature added to do this!

**Comment by jmichaelthurman** [comment]
Yes. I need a feature equivalent to `flux suspend kustomization` to be able to pause healing of applications by automation  that are stood-down during infrastructure maintenance windows.  

For flux apps, I'm using KEDA to launch a job to issue commands & patch scaledObjects in the K8s cluster.

**Comment by crenshaw-dev** [comment]
@jmichaelthurman wouldn't that just be "disable self-heal" and maybe "disable auto-sync"? If you need to prevent even imperative syncs, you could also add a sync window.

**Comment by xpxNPE** [comment]
hi 

another use case :  test/validation environment.

when we are not using our test / validation environment, the "powers that be" delete every application to free up ressources. so when we need the environement again we have to redeploy everything even if there was no change since the last deployment.

having a way to restart the application from the argoCS UI without havaing to redeploy them would be great

**Comment by xpxNPE** [comment]
yes! something that will allow the "power that be" to free ressources when the application are not used, and the devellopers to restart the application easily when needed

**Comment by andrii-korotkov-verkada** [comment]
Are you mostly concerned about Pods' resources. If so, one workaround could be to use KEDA for autoscaling and set annotations
```
autoscaling.keda.sh/paused-replicas: "0"
autoscaling.keda.sh/paused: "true"
```
https://keda.sh/docs/2.16/concepts/scaling-deployments/#pausing-autoscaling

**Comment by marcelldls** [comment]
Our solution has been to add a `removed` and `enabled` key in the app of apps helm chart that we can "patch" from the gui for temporary changes or also commit to git. `Removed` can be used to skip inclusion of a child app (delete all resources). `enabled` can be used to scale the replica count to 0 (suspend without deleting resources). Template: https://github.com/epics-containers/ec-helm-charts/blob/main/Charts/argocd-apps/templates/_apps.tpl. Example use: https://github.com/epics-containers/p47-deployment/blob/main/apps/values.yaml

**Comment by andrii-korotkov-verkada** [comment]
If everything gets deleted while app is suspended, having manifests around in Git is a better strategy, e.g. like @jannfis suggested with appofapps or just in a standalone Git repo, ready to re-apply. I'll close this, but feel free to re-open if these approaches do not work and please explain why.

**Comment by smartaquarius10** [comment]
+1

**Comment by virgiled-bull** [comment]
> How to disable applications, lol.
> 
> Ensure that your ApplicationSet `syncPolicy.automated.prune: true`, then:
> 
> ```
> $ git status
> On branch main
> Your branch is up to date with 'origin/main'.
> 
> Changes to be committed:
>   (use "git restore --staged <file>..." to unstage)
> 	renamed:    applications/matomo/config.json -> applications-disabled/matomo/config.json
> 	renamed:    applications/matomo/values.yaml -> applications-disabled/matomo/values.yaml
> ```
> 
> Note that this may cause gaps in your git history if you move files back and forth. Maybe a smarter person would use symlinks? This is just a joke/hack, but it works. Let's get a real feature added to do this!

Hey, hi, although you say it's a joke, it's the only way I found that seems ok in my context. Problem is I tried it and the application are still showing in the ArgoCD UI. I even tried to delete them, but they keep respawning. Is there something to do so that the application are gone?
(note: I created a disabled directory where I put the application directory with its manifest - I tried also to add the disabled directry in .argocdignore and in the exclude key of the bootstrap application, but still no lick).

**Comment by hyperxpizza** [comment]
:|
